### ISSUE TITLE
licensing: hole_detector.onnx is fine-tuned from Ultralytics (AGPL-3.0) weights and shipped under PolyForm-NC — record provenance and decide; add a licence inventory (incl. Hippocratic react-leaflet)

### ISSUE THREAD

**Issue by cofade** [schutera/highfive#236]
> **Origin:** 2026-08-18 repository audit (findings DEPS-1, DEPS-8); milestone M1 of the audit plan — tracked in #262.

## Summary

The shipped hole-detection model `image-service/models/hole_detector.onnx` (11,081,689 bytes, committed to git and baked into every image-service container) is a YOLO26n-seg fine-tuned from the Ultralytics-distributed pretrained checkpoint `yolo26n-seg.pt`, trained and exported with the `ultralytics` package, which is licensed **AGPL-3.0**. The exported `.onnx` file itself carries Ultralytics metadata declaring `AGPL-3.0 License (https://ultralytics.com/license)`. Ultralytics asserts that models trained from its weights/code are derivative works of the AGPL codebase (this position is legally contested, and the runtime only imports `onnxruntime`, so this is a **judgment call, not a settled fact**). Meanwhile the repo is licensed **PolyForm Noncommercial 1.0.0** and `CONTRIBUTING.md` reserves a paid commercial licence path. Nowhere in the repo — `LICENSE`, `CONTRIBUTING.md`, `image-service/models/README.md`, ADR-027, `docs/` — is the AGPL provenance mentioned (`git grep -in AGPL -- "*.md"` returns nothing; the only hit in the whole tree is inside the binary `.onnx` metadata).

Separately (DEPS-8): `react-leaflet` 5.0.0 and `@react-leaflet/core` 3.0.0 are the only non-standard licence in the npm tree — **Hippocratic-2.1**, a non-OSI use-restriction licence. It is not copyleft and does not conflict with PolyForm-NC, but a project selling commercial licences should have a written third-party licence inventory. There is none.

This issue asks for (1) a provenance record, (2) a **human maintainer decision** on the model's licence posture, recorded in ADR-027, and (3) a small licence inventory table in chapter 02. **No code change until the decision is made.**

## Evidence

Verified against `main@8962c50` on 2026-08-18.

- `dev-tools/ml_hole_detection/train.py:52-55` — the training script's default base weights are the Ultralytics pretrained checkpoint:
  ```py
  ap.add_argument(
      "--model", default="yolo26n-seg.pt",
      help="pretrained weights (yolo26n-seg = latest segment nano; yolo11n-seg = no-upgrade fallback)",
  )
  ```
- `dev-tools/ml_hole_detection/requirements.txt:11` — `ultralytics>=8.3.0` (AGPL-3.0 package; the training + export stack).
- `image-service/models/hole_detector.onnx` — the ONNX metadata embedded in the committed file contains the strings `Ultralytics YOLO26n-seg model trained on C:\Users\...\dev-tools\ml_hole_detection\dataset\holes_seg\data.yaml`, `Ultralytics`, `AGPL-3.0 License (https://ultralytics.com/license)` and `https://docs.ultralytics.com` (extract with `py -c "import re;d=open('image-service/models/hole_detector.onnx','rb').read();[print(m.group().decode()[:200]) for m in re.finditer(rb'[ -~]{8,}',d) if re.search(b'AGPL|licen|ltralytics',m.group(),re.I)]"`). So the artefact self-declares AGPL-3.0, and — privacy nit — bakes in the training machine's absolute local path.
- `image-service/models/README.md:20-27` — "Provenance / how to regenerate" table lists architecture (`YOLO26n-seg`), export, held-out metrics and `Source weights` (`runs/holes_yolo26n_seg/weights/best.pt`, gitignored) — but **no base checkpoint, no licence, no date**.
- `image-service/models/hole_detector.onnx` — 11,081,689 bytes, committed. `image-service/Dockerfile.dev:20` `COPY . .` bakes it into the container; the PM2 production path (`scripts/deploy.sh`) runs from the same checkout.
- `image-service/requirements.txt:24` — `onnxruntime>=1.23.2`; `ultralytics` is **not** a runtime dependency (mitigating: exposure is limited to the weights-as-derivative-work theory).
- `docs/09-architecture-decisions/adr-027-hole-detection-model.md:24-32` (`## Decision`) — "train an Ultralytics **YOLO26n-seg** instance segmenter offline … torch and ultralytics (the training stack) never enter the service image." The ADR frames the `.onnx` purely as a build artifact; no `## Licensing` section exists (sections are Status / Context / Decision / Consequences).
- `LICENSE:1-3` — PolyForm Noncommercial 1.0.0, "HighFive is source-available software, free for noncommercial use".
- `CONTRIBUTING.md:214-218` — "HighFive is **source-available, not open source** … **commercial use requires a separate license** from the maintainers".
- `homepage/package.json:39` — `"react-leaflet": "^5.0.0"`. The lockfile and `node_modules` live at the **repo root** (npm workspace, hoisted): `package-lock.json:4523-4528` pins `react-leaflet` 5.0.0 with `@react-leaflet/core ^3.0.0`, `package-lock.json:1044-1045` pins `@react-leaflet/core` 3.0.0; `node_modules/react-leaflet/package.json:16` and `node_modules/@react-leaflet/core/package.json:16` both declare `"license": "Hippocratic-2.1"`. Rest of the npm tree is MIT/ISC/BSD/Apache/0BSD/CC-BY-4.0 — no GPL/AGPL/LGPL.
- `docs/02-constraints/README.md` — sections are "Technical constraints", "Conventions", "What NOT to commit", "What NOT to do"; no licence inventory.

## Why it matters

- If Ultralytics' derivative-work position holds, `hole_detector.onnx` cannot be redistributed under PolyForm-NC, and a commercial licence sold by the maintainers would **not** cover the detector — the exact artifact that makes the product work. That undercuts the commercial-licence path `CONTRIBUTING.md` advertises.
- Even if the position does not hold, downstream users currently receive no notice of the model's provenance (other than metadata buried in the binary), which is at minimum a due-diligence gap for a project that markets a licensing model.
- Hippocratic-2.1 is rejected by some corporate OSS-intake policies (it is not OSI-approved). A commercial licensee should be able to see it listed instead of discovering it in `node_modules`.

## Proposed change

Docs + one recorded decision; no code until the decision.

1. **Provenance section** in `image-service/models/README.md` (extend the existing "Provenance / how to regenerate" table): base checkpoint (`yolo26n-seg.pt`, Ultralytics pretrained, AGPL-3.0), training stack (`ultralytics>=8.3.0`, `train.py` / `export_onnx.py`), training date, dataset origin (DINO+SAM labels per ADR-027), the fact that the exported file embeds Ultralytics licence metadata, and a pointer to the licence decision in ADR-027.
2. **Maintainer decision**, recorded as a new `## Licensing` section in `docs/09-architecture-decisions/adr-027-hole-detection-model.md` with the decision date and the reasoning. Options to choose from (pick one, and say why the others were rejected):
   - **(a) Treat the `.onnx` as AGPL-3.0**: add an AGPL notice next to the model (e.g. `image-service/models/LICENSE-hole_detector.md` or a paragraph in `LICENSE`), state that the model file is licensed separately from the rest of the repo, and note that a commercial HighFive licence does not relicense the model.
   - **(b) Obtain an Ultralytics Enterprise licence**: record the licence reference/date in the ADR (no secrets/contract text in the repo).
   - **(c) Retrain from a non-AGPL base or from scratch** (e.g. a permissively licensed segmentation backbone; or train YOLO-architecture weights from random init using a non-AGPL training stack). This is the only option that later requires code/model changes — open a follow-up issue for it if chosen; the retrained model still ships via `image-service/models/` per the existing README regeneration steps.
   - **(d) Accept the risk knowingly** (the runtime imports only `onnxruntime`; the derivative-work claim is contested) — still needs a written rationale and a provenance notice.
3. **Licence inventory table** in `docs/02-constraints/README.md` (new "Third-party licences" subsection): project licence (PolyForm-NC 1.0.0), the model provenance row pointing at ADR-027, `react-leaflet` 5 / `@react-leaflet/core` 3 (Hippocratic-2.1 — use-restriction, non-OSI, compatible with noncommercial use, note only), and a one-line statement that the remainder of the npm/pip trees is permissive (MIT/ISC/BSD/Apache/0BSD/CC-BY-4.0). Include the command used to generate the inventory so it can be re-run — from the **repo root** (the npm workspace hoists `node_modules` there):
   ```powershell
   npx license-checker --summary
   py -m pip install pip-licenses
   py -m piplicenses --from=mixed --format=markdown
   ```
   (Verify the tool names before adding them to the doc; do not commit `node_modules` output verbatim.)
4. Add ADR-027 index row wording in `docs/09-architecture-decisions/README.md:47` if the status changes (e.g. "Accepted — licensing addendum 2026-MM-DD").

## Acceptance criteria

- [ ] `docs/09-architecture-decisions/adr-027-hole-detection-model.md` has a `## Licensing` section stating the base checkpoint, the AGPL-3.0 provenance, the chosen option (a/b/c/d), the rejected alternatives, and the decision date, signed off by a human maintainer.
- [ ] `image-service/models/README.md` "Provenance / how to regenerate" table lists base checkpoint, training stack + version, training date, and links to the ADR-027 licensing section.
- [ ] If option (a) is chosen: an AGPL notice ships next to `image-service/models/hole_detector.onnx` and `LICENSE` / `CONTRIBUTING.md` state that the model file is licensed separately.
- [ ] If option (c) is chosen: a follow-up issue exists for the retrain, referencing this one.
- [ ] `docs/02-constraints/README.md` has a "Third-party licences" table listing PolyForm-NC, the model provenance row, and `react-leaflet` / `@react-leaflet/core` (Hippocratic-2.1), plus the regeneration command.
- [ ] `git grep -in "AGPL" -- "docs/*.md" "image-service/models/*.md"` returns at least the ADR-027 and models/README hits.
- [ ] No secrets, contract text, or licence keys are committed.

## Tests

Docs-only; no runtime behaviour changes. Verification commands (PowerShell 5.1, from the repo root):

```powershell
git grep -in "AGPL" -- "*.md"                     # expect hits in ADR-027, image-service/models/README.md, docs/02
git grep -n "Hippocratic" -- "docs/02-constraints/README.md"
git grep -n "## Licensing" -- "docs/09-architecture-decisions/adr-027-hole-detection-model.md"
Get-Content node_modules/react-leaflet/package.json | Select-String '"license"'   # confirms Hippocratic-2.1 still current (root node_modules, npm workspace)
make check-citations
```

If option (c) is later implemented, the existing `image-service` pytest suite (`cd image-service; py -m pytest tests/ -q`) and the ADR-027 export/verify steps in `image-service/models/README.md` apply — but that is the follow-up issue's scope.

## Docs to update

- `docs/09-architecture-decisions/adr-027-hole-detection-model.md` — new `## Licensing` section (decision + date). Per CLAUDE.md "New design decision" row.
- `docs/09-architecture-decisions/README.md` — index row 47 status/notes if changed.
- `image-service/models/README.md` — provenance rows.
- `docs/02-constraints/README.md` — "Third-party licences" inventory table.
- `LICENSE` / `CONTRIBUTING.md` — only if option (a) is chosen (model licensed separately).
- `docs/11-risks-and-technical-debt/README.md` — "Lessons learned" entry: *what*: shipped a fine-tuned model without recording base-weight licence (even though the exported file self-declares AGPL-3.0 in its metadata); *why*: ADR-027 framed the `.onnx` as a build artifact; *how to avoid*: any committed model/weights file must carry a provenance + licence row in its README and in the chapter-02 inventory.

## Effort / risk / dependencies

- **Effort:** S (docs) + one maintainer decision. Option (c), if chosen, is a separate M-sized follow-up.
- **Risk:** None for the docs change. The decision itself is legal/business judgment; the derivative-work claim is Ultralytics' asserted position, not settled law — record it as such.
- **Dependencies:** none. Not blocked by other audit issues.

## Related

- ADR-027 (`docs/09-architecture-decisions/adr-027-hole-detection-model.md`) — model choice and training stack.
- ADR-028 (`docs/09-architecture-decisions/adr-028-ml-inference-server-side-only.md`) — inference is server-side only; the model ships via `image-service`, never in firmware, so this issue has no firmware/OTA component.
- ADR-029 (`image-service/requirements.txt:22-24`) — numpy 2 / onnxruntime pin context.
- #262 — audit epic.

## Guardrails

- **Read-only until the decision.** Do not delete, retrain, or replace `hole_detector.onnx` under this issue; a retrain (option c) is a follow-up with its own PR.
- **Never open a DuckDB connection from `image-service`** and never move inference to the ESP (ADR-001, ADR-028) — irrelevant to the docs change, but a retrain follow-up must respect them.
- **No secrets in the repo**: an Ultralytics Enterprise licence reference goes into the ADR as a date + "held by maintainers", never the licence key or contract.
- **Docs gate**: update the arc42 targets listed under "Docs to update" (ADR-027, chapter 02, chapter 11 lesson) — the PR is not done until they land.
- Run `make check-citations` before review, then the **`senior-reviewer` subagent gate** before opening the PR.
- Use `path's `symbol`` / section anchors over bare `path:line` in the docs you write; line citations drift.
- No auto-close keywords next to `#N` in the PR body — only in the title.
